### PR TITLE
Ab/fixes

### PR DESCRIPTION
**Pull request by AlexandreBossard**
This fix some typo and sentences style. There is a wip on the protocol, but there is more to come.

### CODE REVIEW

File: concepts.md [hunk 1]
```diff
@@ -21,7 +21,7 @@
 [Public Provisional Identity]: concepts.md#public-provisional-identity "Same as Public Permanent Identity, but for a user not registered on the Trustchain yet"
 
 
-*Tanker Core* provides security based on the principle of separation of knowledge between the *Tanker server*, the *user* and the *application server*.
+*Tanker Core* security is based on the separation of knowledge, between the *Tanker server*, the *user*, and the *application server*.
```
Comment (blastrock) [inline concepts.md]: `*Tanker Core*'s security`

File: concepts.md [hunk 2]
```diff
@@ -193,13 +195,12 @@ When received by a *device*, they are stored in the [Local Encrypted Storage].
...
 
-The [Verification Key] is used to register new devices in a secure manner. See the [procotol documentation](protocol.md) for more details.
+The [Verification Key] is used to register new devices in a secure manner. See the [protocol documentation](protocol.md) for more details.
 
 ### Secret Provisional Identity
 
 A [Secret Provisional Identity] represents the identity of a user that is not yet registered on Tanker. It is split into two halves that are stored on the *application server* and on *Tanker server*s.
-A provisional identity is attached to some authentication methods. For the moment only email is supported.
-Each half contains the name of the authentication method, a value (in case of email, the value is the email address), and encryption and signature key pairs.
+Each half contains the name of the authentication method, a value (in case of email, the value is the email address), an encryption and a signature key pairs.
```
Comment (blastrock) [inline concepts.md]: `an encryption key pair, and a signature key pair` or the previous sentence which was correct too

File: concepts.md [hunk 3]
```diff
@@ -193,13 +195,12 @@ When received by a *device*, they are stored in the [Local Encrypted Storage].
 
 ### Verification key
 
-The [Verification Key] is used to register new devices in a secure manner. See the [procotol documentation](protocol.md) for more details.
+The [Verification Key] is used to register new devices in a secure manner. See the [protocol documentation](protocol.md) for more details.
 
 ### Secret Provisional Identity
 
 A [Secret Provisional Identity] represents the identity of a user that is not yet registered on Tanker. It is split into two halves that are stored on the *application server* and on *Tanker server*s.
-A provisional identity is attached to some authentication methods. For the moment only email is supported.
```
Comment (blastrock) [inline concepts.md]: Why remove this?

File: concepts.md [hunk 4]
```diff
@@ -255,7 +255,7 @@ A [Public Provisional Identity] can be generated from a [Secret Provisional Iden
   <dd>a user of the <em>application</em>, owning one or more <em>device</em>s</dd>
 
   <dt>user group</dt>
-  <dd>a group of <em>uer</em>s, the <em>group member</em>s, created and managed by <em>user</em>s. <em>Data</em> shared with a <em>user group</em> is accessible to all its <em>group member</em>s</dd>
+  <dd>a group of <em>user</em>s, the <em>group member</em>s, created and managed by <em>user</em>s. <em>Data</em> shared with a <em>user group</em> is accessible to all its <em>group member</em>s</dd> <-- this weird
```
Comment (blastrock) [inline concepts.md]: `<dd>a group of <em>user</em>s, created and managed by the <em>group members</em>s themselves. <em>Data</em> shared with a <em>user group</em> is accessible to all its <em>group member</em>s</dd>`

File: overview.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # Overview
 
-*Tanker* is a solution for implementing client-side encryption in *application*s. It supports end-to-end encryption, and, in case of *device* loss, it provides an optional way for developers to ensure their *user*s can recover their cryptographic identities. This mechanism is referred to as *identity verification service* and depends on both the *application server* and a third-party verification server to protect and redistribute *user*s' cryptographic identities.
+*Tanker* is a solution for implementing client-side encryption in *application*s. It uses end-to-end encryption, and, in case of *device* loss, it provides an optional way for developers to ensure their *user*s can recover their cryptographic identities. This mechanism is referred to as *identity verification service* and depends on both the *application server* and a third-party verification server to protect and redistribute *user*s' cryptographic identities.
```
Comment (blastrock) [inline overview.md]: makes use of? implements?

File: trustchain_design.md [hunk 6]
```diff
@@ -29,11 +29,11 @@
...
 
-The basic role of a *Trustchain* is to provide public key distribution between *users* without being able to tamper with anything or access any private data.
+The primary role of the *Trustchain* server is to provide public key distribution between *users* without being able to tamper with anything or access any private data. This guaranty is provided by the *Trustchain* itself.
 
 ## Users and devices
 
-Devices are the basic entities we considered when designing the *Trustchain*. A *device* is an entity owned by a *user*. It stores a [Device Signature Key Pair] and a [Device Encryption Key Pair]. In practice, *devices* can be *users*' web browsers, mobile phones, desktop applications, etc.
+Devices are the simplest entities we considered when designing the *Trustchain*. A *device* is an entity owned by a *user*. It stores a [Device Signature Key Pair] and a [Device Encryption Key Pair]. In practice, *devices* can be *users*' web browsers, mobile phones, desktop applications, etc.
```
Comment (blastrock) [inline trustchain_design.md:36]: it's not really simple, a key publish is simple. basic was more appropriate here imho

File: trustchain_design.md [hunk 7]
```diff
@@ -112,25 +112,25 @@ The author also stores the block they just emitted, preventing the *Trustchain*
...
 
 ### Forward secrecy
 
-Forward secrecy is, despite being an interesting property, **not **implemented in *Tanker Core*. The reason is that *Tanker Core* is not targeted solely at encrypting ephemeral message exchanges, but also persistent data. We want *users* to be able to decrypt anything at any time with any of their *devices*, which means that we cannot use a forward secure scheme.
+Forward secrecy is, despite being an interesting property, **not** implemented in *Tanker Core*. The reason is that *Tanker Core* is not targeted solely at encrypting ephemeral message exchanges, but also persistent data. We want *users* to be able to decrypt anything at any time with any of their *devices*, which means that we cannot use a forward secure scheme.
 
-We consider implementing an optional forward secure scheme tailored for messaging in a future version of *Tanker Core*.
+We are considering implementing an optional forward secure scheme tailored for messaging in a future version of *Tanker Core*.
```
Comment (blastrock) [inline trustchain_design.md]: delete this? this is not an advertisement flyer

File: trustchain_design.md [hunk 8]
```diff
@@ -112,25 +112,25 @@ The author also stores the block they just emitted, preventing the *Trustchain*
...
+*Tanker Core* is based on the separation of trust principle. Any *user* needs to trust:
 
 - the *application* to store and distribute encrypted data
 - *Tanker* to distribute blocks
 - *Tanker* and the *application* not to collude
 
 A collusion between *Tanker* and the *application* could result in a split view attack, in which an alternate, seemingly valid version of the *Trustchain* is strategically distributed to *users* in order to trick them into sharing data with the wrong entity.
 
-We plan on implementing a gossiping scheme in future versions of *Tanker Core*, which would prevent split view attacks.
+We are considering to implement a gossiping scheme in future versions of *Tanker Core*, which would prevent split view attacks.
```
Comment (blastrock) [inline trustchain_design.md]: implementing*

https://thegrammarexchange.infopop.cc/topic/consider-to-do-consider-doing

remove this too?

File: concepts.md [hunk 9]
```diff
@@ -200,6 +200,7 @@ The [Verification Key] is used to register new devices in a secure manner. See t
 ### Secret Provisional Identity
 
 A [Secret Provisional Identity] represents the identity of a user that is not yet registered on Tanker. It is split into two halves that are stored on the *application server* and on *Tanker server*s.
+A provisional identity is attached to some authentication methods. For the moment, email and OpenID connect IDToken are supported.
```
Comment (blastrock) [inline concepts.md]: tokens

File: concepts.md [hunk 10]
```diff
@@ -207,6 +207,16 @@ Each half contains the name of the authentication method, a value (in case of em
 
 A [Public Provisional Identity] can be generated from a [Secret Provisional Identity] and consists of the [Secret Provisional Identity] without its secrets parts.
 
+### TLS
+
+Transport Layer Security is always used between *Tanker Core* and *Tanker Server* to exchange information.
+This guarantee secrecy of the data exchanged.
```
Comment (blastrock) [inline concepts.md]: `guarantees the confidentiality and authenticity`

File: concepts.md [hunk 11]
```diff
@@ -207,6 +207,16 @@ Each half contains the name of the authentication method, a value (in case of em
...
+
+### Resource ID
+
+A randomly generated unique ID part of an encrypted data.
```
Comment (blastrock) [inline concepts.md]: it is not always random

File: protocol.md [hunk 12]
```diff
@@ -110,13 +110,11 @@ If any of these check fail, the connection is closed.
...
 Prerequisite: the *user* has created a [Verification Key].
 
-Except for the first *device*, which is validated in a specific way previously described, additional *device*s must be validated by an already registered *device*. In practice, these *device*s are validated by the *ghost device*.
+Except for the first *device*, which is validated in a specific way previously described in the [Sign up process](#sign-up), additional *device*s must be validated by an already registered *device*. In practice, these *device*s are validated by the *virtual device*.
 
 Given the *user*'s [Verification Key], the steps to register a new *device* are as follows:
 
-1. Extract the *ghost device*'s [Device Encryption Key Pair] and [Device Signature Key Pair] from the [Verification Key]
-2. Pull the *Trustchain* up to the *ghost device*'s `device_creation` *block*, verify it and extract the [User Encryption Key Pair] from it
-3. Decrypt the [User Encryption Key Pair] using the  *ghost device*'s private [Device Encryption Key Pair]
```
Comment (blastrock) [inline protocol.md:119]: you can't skip this, you need the user encryption key pair to make the device block

File: protocol.md [hunk 13]
```diff
@@ -29,6 +31,9 @@ The protocols have been split in 3 sections for ease of reading:
 - The [encryption](#encryption) section describes how *resource*s are encrypted and shared with *user*s
 - The [group encryption](#group-encryption) section describes how *user group*s are managed and how to share encrypted *resource*s with them
 
+# Communication
+
+Every communication and information exchanges between *Tanker Core* and the *Tanker Server* are done through [TLS] connection. In the same way, exchanges between *Tanker Core* and the *application server* also use [TLS], in particular for Tanker Identity retrievals.
```
Comment (blastrock) [inline protocol.md:36]: this is the third time you say this

File: protocol.md [hunk 14]
```diff
@@ -50,21 +55,21 @@ What happens next depends on the chosen method:
...
 If the user chose 'email' or 'passphrase' as the verification method, the [Verification Key] is encrypted on the *device* with the [User Secret] and sent to the *Tanker server*.
 
 #### Signing in
 
-The first time the user signs in on a new device, their identity must be *verified* using the method selected during sign up.
+The first time the user signs in on a new device, their identity must be *verified* using the method selected during the sign up process.
 
 * By passphrase: the *user* must provide their passphrase to *Tanker Core*. It is hashed client-side, then sent to the *Tanker server* to fetch the encrypted [Verification Key].
-* By email: the *user* triggers a verification request through to the *application server* which in turn calls the *Tanker server*. The *Tanker server* then sends an email containing the *Tanker verification code* that must be provided to *Tanker Core*.
+* By email: the *user* triggers a verification request through the *application server* which in turn calls the *Tanker server*. Then, the *Tanker server* sends an email containing the *Tanker verification code* which must be provided to *Tanker Core*.
```
Comment (blastrock) [inline protocol.md:69]: by the user

File: protocol.md [hunk 15]
```diff
@@ -50,21 +55,21 @@ What happens next depends on the chosen method:
...
 * By verification key: the *user* directly gives their verification key to *Tanker Core*.
 
-When using the 'passphrase', or 'email' verification methods, the *user* still needs to authenticate against the *application server* to obtain their [User Secret] to decrypt the [Verification Key].
+While using the 'passphrase', or 'email' verification methods, the *user* still needs their [User Secret] to decrypt the [Verification Key]. They would need to authenticate against the *application server* to obtain it.
```
Comment (blastrock) [inline protocol.md:72]: `When` was better. Why `would`?

File: protocol.md [hunk 16]
```diff
@@ -115,41 +120,51 @@ Except for the first *device*, which is validated in a specific way previously d
...
+## Encryption and Decryption
 
-### Data encryption and decryption
+### Data encryption
 
 Prerequisite: the *user*'s *device* is authenticated against the *Tanker server*.
 
-Encrypting *data* implies automatically sharing the [Resource Encryption Key] with the user themselves so that their other devices can decrypt the data. The steps to encrypt a resource are as follows:
+Encrypting *data* implies automatically sharing the [Resource Encryption Key] with the user themselves so that their other devices can decrypt the data.
+Recipients, users and/or groups, can be provided so that the [Resource Encryption Key] be shared to them.
```
Comment (blastrock) [inline protocol.md]: so that the resource encryption key can* be shared with* them

File: protocol.md [hunk 17]
```diff
@@ -115,41 +120,51 @@ Except for the first *device*, which is validated in a specific way previously d
...
-5. *Tanker Core* asymmetrically encrypts the [Resource Encryption Key] with the public [User Encryption Key Pair], creating the [Shared Encrypted Key]
-6. *Tanker Core* creates a `key_publish` *block* containing the [Shared Encrypted Key] and the recipient's public [User Encryption Key Pair], and pushes it to the *Trustchain*
-7. The *Tanker server* validates the *block* and notifies all the recipients' *device*s
-8. The recipient's *device* retrieves the *block*, verifies it and decrypts the [Shared Encrypted Key] using the private [User Encryption Key Pair], obtaining the [Resource Encryption Key]
-9. The *application* calls `tanker.decrypt` on the recipient's *device*
-10. *Tanker Core* decrypts the data using the [Resource Encryption Key]
+1. The *application* fetches the recipients' [Public Permanent Identity] from the *application server*
+1. *Tanker Core* fetches the recipients' `device_creation` *block*s from the *Trustchain*
+1. *Tanker Core* verifies the received `device_creation` *block*s
+1. *Tanker Core* extracts the recipient's public [User Encryption Key Pair]
```
Comment (blastrock) [inline protocol.md:154]: it's not a pair if you use only the public key (same below, multiple occurrences)

File: protocol.md [hunk 18]
```diff
@@ -115,41 +120,51 @@ Except for the first *device*, which is validated in a specific way previously d
...
+1. For each recipient, *Tanker Core* creates a `key_publish` *block* containing the [Shared Encrypted Key] and the recipient's public [User Encryption Key Pair], and pushes it to the *Trustchain*
+1. The *Tanker server* validates the *block*
+
+## Data decryption
+
+Prerequisite: the user's has access to the encrypted data
+
+1. *Tanker Core* extracts the [Resource ID] from the encrypted data
+1. *Tanker Core* retrieves for the recipients' *device* the `key_publish` *block*,
+1. *Tanker Core* verifies the retrieved block
```
Comment (blastrock) [inline protocol.md:165]: not anymore, we should document that btw in block verification, and maybe say again here that we do *not* verify these blocks

File: protocol.md [hunk 19]
```diff
@@ -159,33 +174,41 @@ Prerequisite: the *user*'s *device* is authenticated against the *Tanker server*
...
-11. Each *group member*s' *device* retrieves the *block*, verifies it and decrypts the [Group Encryption Key Pair] and [Group Signature Key Pair] using their private [User Encryption Key Pair]
+1. *Tanker Core* fetches all future *group member*s' `device_creation` *block*s from the *Trustchain*
+1. *Tanker Core* verifies them and extracts their public [User Encryption Key Pair]
+1. *Tanker Core* generates the [Group Encryption Key Pair] and the [Group Signature Key Pair]
+1. *Tanker Core* encrypts the private [Group Signature Key Pair] with the public [Group Encryption Key Pair]
+1. *Tanker Core* encrypts the private [Group Encryption Key Pair] with each future *group member*'s public [User Encryption Key Pair]
+1. Using the private [Group Signature Key Pair], *Tanker Core* signs the public and encrypted private [Group Encryption Key Pair] and [Group Signature Key Pair]
+1. *Tanker Core* creates a `user_group_creation` *block* with all of the above and pushes it to the *Trustchain*
+1. The *Tanker server* validates the *block* and sends the update to all *group member*s' *device*s
```
Comment (blastrock) [inline protocol.md:184]: skip the "sends the update..." part? it's wrong for js, and soon for c++ too

File: protocol.md [hunk 20]
```diff
@@ -159,33 +174,41 @@ Prerequisite: the *user*'s *device* is authenticated against the *Tanker server*
...
-2. *Tanker Core* fetches the recipient *group*'s public [Group Encryption Key Pair] from the *Trustchain*, if not already present in the [Local Encrypted Storage]
-3. *Tanker Core* encrypts the [Resource Encryption Key] with the public [Group Encryption Key Pair]. The result is the [Shared Encrypted Key] for this particular *user group* and *resource*
-4. *Tanker Core* creates a *block* containing the [Shared Encrypted Key] and the recipient public [Group Encryption Key Pair]
-5. *Tanker Core* pushes the *block* to the *Trustchain*
-6. The *Tanker server* validates the *block* and notifies all the recipients' *device*s
-8. The recipients' *device*s retrieve the *block*, verify it and decrypt the [Shared Encrypted Key] using the private [Group Encryption Key Pair], obtaining the [Resource Encryption Key]
-7. The *application* calls `tanker.decrypt` on one of the recipient group's *device*s
-10. *Tanker Core* decrypts the data using the [Resource Encryption Key]
+1. The *application* fetches the [GID](#group-id) to share with
+1. *Tanker Core* look for the [Group Encryption Key Pair] associated with [GID](#group-id) in the [Local Encrypted Storage]
```
Comment (blastrock) [inline protocol.md:195]: This is the first time we talk about looking things up in the local cache before doing stuff, but we do it for users and resource keys too. We should say that everywhere, or nowhere, but not just here.

also looks*

File: protocol.md [hunk 21]
```diff
@@ -159,33 +174,41 @@ Prerequisite: the *user*'s *device* is authenticated against the *Tanker server*
...
+1. *Tanker Core* fetches all future *group member*s' `device_creation` *block*s from the *Trustchain*
+1. *Tanker Core* verifies them and extracts their public [User Encryption Key Pair]
+1. *Tanker Core* generates the [Group Encryption Key Pair] and the [Group Signature Key Pair]
+1. *Tanker Core* encrypts the private [Group Signature Key Pair] with the public [Group Encryption Key Pair]
+1. *Tanker Core* encrypts the private [Group Encryption Key Pair] with each future *group member*'s public [User Encryption Key Pair]
+1. Using the private [Group Signature Key Pair], *Tanker Core* signs the public and encrypted private [Group Encryption Key Pair] and [Group Signature Key Pair]
+1. *Tanker Core* creates a `user_group_creation` *block* with all of the above and pushes it to the *Trustchain*
+1. The *Tanker server* validates the *block* and sends the update to all *group member*s' *device*s
+1. Each *group member*s' *device* retrieves the *block*, verifies it and decrypts the [Group Encryption Key Pair] and [Group Signature Key Pair] using their private [User Encryption Key Pair]
```
Comment (blastrock) [inline protocol.md:185]: you can move that to the decrypting part

File: protocol.md [hunk 22]
```diff
@@ -159,33 +174,41 @@ Prerequisite: the *user*'s *device* is authenticated against the *Tanker server*
...
+1. *Tanker Core* pushes the *block* to the *Trustchain*
+1. The *Tanker server* validates the *block*
+
+### Decrypting with user groups
+
+Prerequisite: the *user*'s *device* is authenticated against the *Tanker server*, they have access to the encrypted data and they are member of the current Tanker group.
+
+1. *Tanker Core* extracts the [Resource ID] from the encrypted data
+1. *Tanker Core* retrieves for the recipients' *device* the `key_publish` *block*,
+1. *Tanker Core* verifies the retrieved block
```
Comment (blastrock) [inline protocol.md:209]: nope

**Review by blastrock — CHANGES_REQUESTED** [review]

File: concepts.md (hunk 2 above)
Comment (AlexandreBossard) [reply concepts.md]: Not sure what difference it makes...

File: trustchain_design.md (hunk 6 above)
Comment (AlexandreBossard) [reply trustchain_design.md:36]: `first` works too. I don't like `basic`, it's not the right adjectives in this context.

File: trustchain_design.md (hunk 7 above)
Comment (AlexandreBossard) [reply trustchain_design.md]: yes

File: trustchain_design.md (hunk 8 above)
Comment (AlexandreBossard) [reply trustchain_design.md]: yes too

File: concepts.md (hunk 10 above)
Comment (AlexandreBossard) [reply concepts.md]: go fix wikipedia!

File: concepts.md (hunk 11 above)
Comment (AlexandreBossard) [reply concepts.md]: You talked about stream i supposed. But stream ID are not resource ID, right ?

File: protocol.md (hunk 20 above)
Comment (AlexandreBossard) [reply protocol.md:195]: Question is, does it have an impact on the protocol security, like the on the demand ?

**Review by blastrock — APPROVED** [review]

**Pull request merged**